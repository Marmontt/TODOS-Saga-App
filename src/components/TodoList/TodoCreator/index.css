.todoCreator-container {
    padding: 1em;
    margin-top: 10px;
    height: 20px;
    display: flex;
    align-items: center;
    border-radius: 5px;
}

.todoCreator-content {
    display: flex;
    align-items: center;
    position: fixed;
}

.todoCreator-line {
    flex-grow: 1;
    height: 1px;
}

.todoCreator-text {
    color: white;
    padding-left: 10px;
    font-family: Roboto, serif;
    font-weight: 500;
}

.todoCreator-checkbox {
    width: 1.5em;
    height: 1.5em;
    border-right: 1px solid #fff;
    z-index: 1;
}