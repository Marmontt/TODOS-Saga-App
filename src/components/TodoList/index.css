.todoList-container {
    margin: 0 auto 1em auto;
    width: 60vw;
    padding-top: 10px;
}

span {
    user-select: none;
}

.todoList-title {
    font-family: Roboto, serif;
    font-weight: bold;
    color: #555555;
}

.todoList-fulfilled {
    font-family: Roboto, serif;
    font-weight: 500;
    color: #555555;
    padding-right: 15px;
}

.todoList-menu {
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
}

.todoList-fulfilled-container {
    display: flex;
    align-items: center;
}

.switch {
    position: relative;
    width: 50px;
    height: 26px;
}

.switch input {
    opacity: 0;
    width: 0;
    height: 0;
}

.slider {
    position: absolute;
    display: flex;
    align-items: center;
    cursor: pointer;
    top: 0;
    left: -5px;
    right: 0;
    bottom: 0;
    background-color: #ccc;
    -webkit-transition: .4s;
    transition: .4s;
    border: 2px solid #5f5f5f;
}

.slider:before {
    position: absolute;
    content: "";
    height: 16px;
    width: 16px;
    left: 4px;
    background-color: #5f5f5f;
    -webkit-transition: .4s;
    transition: .4s;
}

input:checked + .slider {
    background-color: #eb716a;
}

input:focus + .slider {
    box-shadow: 0 0 1px #eb716a;
}

input:checked + .slider:before {
    -webkit-transform: translateX(26px);
    -ms-transform: translateX(26px);
    transform: translateX(26px);
}

/* Rounded sliders */
.slider.round {
    border-radius: 15px;
}

.slider.round:before {
    border-radius: 50%;
}
