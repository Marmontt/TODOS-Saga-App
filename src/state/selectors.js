export const getAllTodos = (state) => state.todos.todos;

export const getTodos = (state, fulfilled = false) => {
    if (!fulfilled) {
        return state.todos.todos.filter(todo => todo.fulfilled === false)
    } else {
        return state.todos.todos
    }
};
