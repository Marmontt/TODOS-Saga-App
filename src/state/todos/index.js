import * as todosActions from './action';
import * as todosTypes from './types';

import todoSagas from './sagas';
import todosReducer from './reducer';

export {
    todosActions,
    todosTypes,
    todoSagas,
    todosReducer
}