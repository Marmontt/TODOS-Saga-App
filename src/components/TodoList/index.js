import React, {useState} from 'react';
import TodoCreator from "./TodoCreator";
import './index.css';
import {getAllTodos} from "../../state/selectors";
import {useSelector} from 'react-redux';


const TodoList = () => {
    const todosList = useSelector(getAllTodos);
    const [toggleState, setToggleState] = useState(false);

    return (
        <div className={'todoList-container'}>
            <div className={'todoList-menu'}>
                <span className={'todoList-title'}>React Todos List</span>
                <div className={'todoList-fulfilled-container'}>
                    <span className={'todoList-fulfilled'}>Show fulfilled todos:</span>
                    <label className={'switch'}>
                        <input type={'checkbox'} onChange={() => setToggleState(!toggleState)} checked={toggleState}/>
                        <span className={'slider round'}/>
                    </label>
                </div>
            </div>
            {
                toggleState === false ?
                    todosList
                        .map(({text, color, fulfilled}, index) =>
                            <TodoCreator todoIndex={index} key={index} color={color} text={text}
                                         fulfilled={fulfilled}/>)
                        .filter(todoCreator => todoCreator.props.fulfilled === false) :
                    todosList
                        .map(({text, color, fulfilled}, index) =>
                            <TodoCreator todoIndex={index} key={index} color={color} text={text}
                                         fulfilled={fulfilled}/>)
            }
        </div>
    );
};


export default TodoList;