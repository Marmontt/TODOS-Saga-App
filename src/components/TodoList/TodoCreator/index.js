import React from 'react';
import {useDispatch} from 'react-redux';

import * as todoActions from "../../../state/todos/action";

import './index.css';


const TodoCreator = ({color, text, fulfilled, todoIndex}) => {
    const dispatch = useDispatch();
    return (
        <div style={{background: color}} className={'todoCreator-container'}>
            {fulfilled === true ?
                <hr className={'todoCreator-line'} color={'#000'}/>
                : null}
            <div className={'todoCreator-content'}>
                <input onChange={() => {
                    dispatch(todoActions.setFulfilledState(todoIndex, !fulfilled));
                }} className={'todoCreator-checkbox'} checked={fulfilled} type="checkbox"/>
                <div className={'todoCreator-text'}>{text}</div>
            </div>
        </div>
    );
};


export default TodoCreator;
