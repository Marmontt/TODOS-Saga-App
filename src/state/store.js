import {applyMiddleware, combineReducers, createStore} from 'redux';
import createSagaMiddleware from 'redux-saga';
import {composeWithDevTools} from 'redux-devtools-extension';

import {todosReducer} from "./todos";
import {todoSagas} from './todos'

const rootReducer = combineReducers({
    todos: todosReducer
});

const rootSaga = createSagaMiddleware();

const store = createStore(
    rootReducer,
    composeWithDevTools(applyMiddleware(rootSaga))
);

rootSaga.run(todoSagas);

export default store;
