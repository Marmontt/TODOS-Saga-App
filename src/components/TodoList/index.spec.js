import React from "react";
import {shallow} from "enzyme";

import {TodoList} from "../TodoList";
import {getTodos} from '../../state/selectors';

jest.mock('react-redux', () => ({
    useSelector: jest.fn(fn => fn()),
}));
jest.mock('../../state/selectors');

describe("TodoList", () => {
    it("should render properly", () => {
        getTodos.mockReturnValue([{text: 'asdasd', color: '#fff', fulfilled: false, id: 1}]);
        const wrapper = shallow(<TodoList/>);
        expect(wrapper).toMatchSnapshot();
    });
    it("should respond to toggle change", () => {
        const wrapper = shallow(<TodoList/>);
        wrapper.find("input").simulate("change", {target: true});
        expect(wrapper.find("input")).toHaveProp({checked: true});
    });
});
