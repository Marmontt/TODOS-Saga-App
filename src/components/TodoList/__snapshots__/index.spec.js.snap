// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList should render properly 1`] = `
<div
  className="todoList-container"
>
  <div
    className="todoList-menu"
  >
    <span
      className="todoList-title"
    >
      React Todos List
    </span>
    <div
      className="todoList-fulfilled-container"
    >
      <span
        className="todoList-fulfilled"
      >
        Show fulfilled todos too:
      </span>
      <label
        className="switch"
      >
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="slider round"
        />
      </label>
    </div>
  </div>
  <TodoCreator
    color="#fff"
    fulfilled={false}
    key="0"
    text="asdasd"
    todoIndex={1}
  />
</div>
`;
